File: restty-app/frontend/src/main/frontend/src/app/services/test-case-settings.service.ts
import { IdNameMethod } from '../model/id-name-method';
import { TestCaseSettings } from '../model/test-case-settings';
import { TestCaseService } from './test-case.service';
import { HttpHeaders } from '@angular/common/http';
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';

const httpOptions = {
    headers: new HttpHeaders({ 'Content-Type': 'application/json' })
};

@Injectable()
export class TestCaseSettingsService {

  public static get TEST_CASE_SETTINGS_PATH(): string { return '/settings'; }
  public static get TEST_CASE_STEPS_PATH(): string { return TestCaseSettingsService.TEST_CASE_SETTINGS_PATH + '/steps'; }

  constructor(private http: HttpClient) {}

  /**
   * Finds all test case settings for given test case.
   *
   * @param testCaseId ID of test case to search by
   * @return list of test case settings
   */
  findAllByTestCaseId(testCaseId: number) {
    return this.http.get<TestCaseSettings[]>(
      TestCaseService.TEST_CASE_PATH + `/${testCaseId}` + TestCaseSettingsService.TEST_CASE_SETTINGS_PATH);
  }

  /**
   * Finds test case's settings by given id
   *
   * @param testCaseSettingsId ID of test case settings to search by
   * @return TestCaseSettings
   */
  findById(testCaseSettingsId: number) {
    return this.http.get<TestCaseSettings>('/api/test-cases/settings' + `/${testCaseSettingsId}`);
  }

  /**
   * Finds all steps for given test case.
   *
   * @param testCaseId
   *            ID of test case to search by
   * @return list of step's ids, names and methods
   */
  findAllSteps(testCaseId: number) {
    return this.http.get<IdNameMethod[]>(
      TestCaseService.TEST_CASE_PATH + `/${testCaseId}` + TestCaseSettingsService.TEST_CASE_STEPS_PATH);
  }

  /**
   * Adds endpoint to the test case.
   *
   * @param testCaseId ID of test case
   * @param endpointId ID of endpoint
   * @param usePrevious if the endpoint should be connected to previous step
   * @return testCaseSettings
   */
  addEndpointToTestCase(testCaseId: number, endpointId: number, usePrevious: boolean) {
    return this.http.post<TestCaseSettings>(TestCaseService.TEST_CASE_PATH + `/${testCaseId}`
      + TestCaseSettingsService.TEST_CASE_SETTINGS_PATH + `/endpoints/${endpointId}`, JSON.stringify(usePrevious), httpOptions);
  }

  /**
   * Adds test case to the test case.
   *
   * @param testCaseId ID of test case
   * @param stepTestCaseId ID of step test case
   * @param usePrevious if the endpoint should be connected to previous step
   * @return testCaseSettings
   */
  addTestStepToTestCase(testCaseId: number, testStepId: number, usePrevious: boolean) {
    return this.http.post<TestCaseSettings>(TestCaseService.TEST_CASE_PATH + `/${testCaseId}`
      + TestCaseSettingsService.TEST_CASE_SETTINGS_PATH + `/test-cases/${testStepId}`, JSON.stringify(usePrevious), httpOptions);
  }

  /**
   * Deletes given and following steps
   *
   * @param testCaseId ID of test case
   * @param stepOrder step to delete
   * @return http status
   */
  deleteSteps(testCaseId: number, stepOrder: number) {
    return this.http.request('delete',
      TestCaseService.TEST_CASE_PATH + `/${testCaseId}` + TestCaseSettingsService.TEST_CASE_SETTINGS_PATH, {
        headers: httpOptions.headers,
        body: stepOrder
      });
  }

}

File: restty-app/backend/src/main/java/cz/restty/app/service/TestCaseSettingsService.java
package cz.restty.app.service;

import java.util.List;

import cz.restty.app.entities.Endpoint;
import cz.restty.app.entities.TestCase;
import cz.restty.app.entities.TestCaseSettings;
import cz.restty.app.rest.dto.IdNameMethodDto;

/**
 * Service layer for manipulation with {@link TestCaseSettings} entities.
 * 
 * @author Ondrej Krpec
 *
 */
public interface TestCaseSettingsService {

    /**
     * Finds all steps for given test case.
     * 
     * @param testCase
     *            Test case to search by
     * @return list of step's ids, names and methods
     */
    List<IdNameMethodDto> findAllSteps(TestCase testCase);

    /**
     * Adds endpoint as a step of test case.
     * 
     * @param parent
     *            {@link TestCase}
     * @param step
     *            {@link Endpoint}
     * @param usePrevious
     *            connect the endpoint's input to previous step's output
     * @return {@link TestCaseSettings}
     */
    TestCaseSettings addEndpoint(TestCase parent, Endpoint step, Boolean usePrevious);

    /**
     * Adds test case as a step of test case.
     * 
     * @param parent
     *            {@link TestCase}
     * @param step
     *            {@link TestCase}
     * @return {@link TestCaseSettings}
     */
    TestCaseSettings addTestCase(TestCase parent, TestCase step);

    /**
     * Removes step with given order (and all following steps) from the test case.
     * 
     * @param testCase
     *            {@link TestCase}
     * @param stepOrder
     *            Step's order
     */
    void removeSteps(TestCase testCase, Integer stepOrder);

    /**
     * Removes all steps for given test case
     * 
     * @param testCase
     *            {@link TestCase}
     */
    void removeAllByTestCase(TestCase testCase);

}
